### PR TITLE
Initial commit of audio/I2S driver without of any BSP support for any…

### PR DESCRIPTION
**Pull request by btheosam**
… hardware

### CODE REVIEW

File: os/drivers/audio/Kconfig [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#
+# For a description of the syntax of this configuration file,
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+config AUDIO_I2SCHAR
+	bool "I2S character driver (for testing only)"
+	default n
+	depends on I2S && AUDIO
```
Comment (sunghan-chang) [inline os/drivers/audio/Kconfig]: Where is AUDIO on Kconfig?

File: os/include/tinyara/audio/i2s.h [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+ * Input Parameters:
+ *   dev  - Device-specific state data
+ *   rate - The I2S sample rate in samples (not bits) per second
+ *
+ * Returned Value:
+ *   Returns the resulting bitrate
+ *
+ ****************************************************************************/
+
+#define I2S_RXSAMPLERATE(d, f) ((d)->ops->i2s_rxsamplerate(d,r))
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/i2s.h]: d,r -> d, r

File: os/include/tinyara/audio/i2s.h [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+ * Input Parameters:
+ *   dev   - Device-specific state data
+ *   width - The I2S data with in bits.
+ *
+ * Returned Value:
+ *   Returns the resulting bitrate
+ *
+ ****************************************************************************/
+
+#define I2S_RXDATAWIDTH(d, b) ((d)->ops->i2s_rxdatawidth(d,b))
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/i2s.h]: same as previous

File: os/include/tinyara/audio/i2s.h [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+ *   only enqueues the transfer and returns immediately.  Success here only
+ *   means that the transfer was enqueued correctly.
+ *
+ *   When the transfer is complete, a 'result' value will be provided as
+ *   an argument to the callback function that will indicate if the transfer
+ *   failed.
+ *
+ ****************************************************************************/
+
+#define I2S_RECEIVE(d, b, c, a, t) ((d)->ops->i2s_receive(d,b,c,a,t))
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/i2s.h]: here also

File: os/include/tinyara/audio/i2s.h [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+ *   only enqueues the transfer and returns immediately.  Success here only
+ *   means that the transfer was enqueued correctly.
+ *
+ *   When the transfer is complete, a 'result' value will be provided as
+ *   an argument to the callback function that will indicate if the transfer
+ *   failed.
+ *
+ ****************************************************************************/
+
+#define I2S_SEND(d, b, c, a, t) ((d)->ops->i2s_send(d,b,c,a,t))
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/i2s.h]: ,

File: os/include/tinyara/audio/i2s.h [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+ * Input Parameters:
+ *   dev   - Device-specific state data
+ *   width - The I2S data with in bits.
+ *
+ * Returned Value:
+ *   Returns the resulting bitrate
+ *
+ ****************************************************************************/
+
+#define I2S_TXDATAWIDTH(d, b) ((d)->ops->i2s_txdatawidth(d,b))
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/i2s.h]: ,

File: os/include/tinyara/audio/i2s.h [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+ * Input Parameters:
+ *   dev  - Device-specific state data
+ *   rate - The I2S sample rate in samples (not bits) per second
+ *
+ * Returned Value:
+ *   Returns the resulting bitrate
+ *
+ ****************************************************************************/
+
+#define I2S_TXSAMPLERATE(d, f) ((d)->ops->i2s_txsamplerate(d,r))
```
Comment (sunghan-chang) [inline os/include/tinyara/audio/i2s.h]: ,

File: os/drivers/audio/tone.c [hunk 8]
```diff
@@ -0,0 +1,943 @@
...
+#include <fcntl.h>
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/kmalloc.h>
+#include <tinyara/fs/fs.h>
+#include <tinyara/arch.h>
+#include <tinyara/drivers/pwm.h>
+#include <tinyara/timers/oneshot.h>
```
Comment (sunghan-chang) [inline os/drivers/audio/tone.c]: We don't have this header

File: os/drivers/audio/tone.c [hunk 9]
```diff
@@ -0,0 +1,943 @@
...
+#include <semaphore.h>
+#include <fcntl.h>
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/kmalloc.h>
+#include <tinyara/fs/fs.h>
+#include <tinyara/arch.h>
+#include <tinyara/drivers/pwm.h>
```
Comment (sunghan-chang) [inline os/drivers/audio/tone.c]: Is it same as tinyara/pwm.h? We don't have tinyara/drivers/pwm.h.

File: os/drivers/audio/Kconfig (hunk 1 above)
Comment (btheosam) [inline os/drivers/audio/Kconfig]: Will be part of Framework layer soon.

**Comment by btheosam** [comment]
this branch got corrupted after pull

**Pull request closed without merging**